app/scripts/controllers/threebox.js

```diff
@@ -0,0 +1,181 @@
+const providerFromEngine = require('eth-json-rpc-middleware/providerFromEngine')
+const createMetamaskMiddleware = require('./network/createMetamaskMiddleware')
+const createOriginMiddleware = require('../lib/createOriginMiddleware')
+
+class ThreeBoxController {
+  constructor (opts = {}) {
+    const {
+      preferencesController,
+      keyringController,
+      addressBookController,
+      version,
+      getKeyringControllerState,
+      getSelectedAddress,
+      signPersonalMessage,
+    } = opts
+
+    this.preferencesController = preferencesController
+    this.addressBookController = addressBookController
+    this.keyringController = keyringController
+    this.provider = this._createProvider({
+      static: {
+        eth_syncing: false,
+        web3_clientVersion: `MetaMask/v${version}`,
+      },
+      version,
+      getAccounts: async ({ origin }) => {
+        if (origin !== '3Box') { return [] }
+        const isUnlocked = getKeyringControllerState().isUnlocked
+
+        const selectedAddress = getSelectedAddress()
+
+        if (isUnlocked && selectedAddress) {
+          return [selectedAddress]
+        } else {
+          return []
+        }
+      },
+      processPersonalMessage: (msgParams) => {
+        return Promise.resolve(signPersonalMessage(msgParams))
+      },
+    })
+
+    const initState = {
+      threeBoxSyncingAllowed: true,
+      restoredFromThreeBox: null,
+      ...opts.initState,
+      threeBoxAddress: null,
+      threeBoxSynced: false,
+    }
+    this.store = new ObservableStore(initState)
+    this.registeringUpdates = false
+
+    this.init()
+  }
+
+  async init () {
+    const accounts = await this.keyringController.getAccounts()
+    this.address = accounts[0]
+    if (this.address && !(this.box && this.store.getState().threeBoxSynced)) {
+      await this.new3Box(this.address)
+    }
+  }
+
+  async _update3Box ({ type }, newState) {
+    const { threeBoxSyncingAllowed, threeBoxSynced } = this.store.getState()
+    if (threeBoxSyncingAllowed && threeBoxSynced) {
+      await this.space.private.set('lastUpdated', Date.now())
+      await this.space.private.set(type, JSON.stringify(newState))
+    }
+  }
+
+  _createProvider (providerOpts) {
+    const metamaskMiddleware = createMetamaskMiddleware(providerOpts)
+    const engine = new JsonRpcEngine()
+    engine.push(createOriginMiddleware({ origin: '3Box' }))
+    engine.push(metamaskMiddleware)
+    const provider = providerFromEngine(engine)
+    return provider
+  }
+
+  _waitForOnSyncDone () {
+    return new Promise((resolve) => {
+      this.box.onSyncDone(() => {
+        log.debug('3Box box sync done')
+        return resolve()
+      })
+    })
+  }
+
+  async new3Box (address) {
+    if (this.getThreeBoxSyncingState()) {
+      this.store.updateState({ threeBoxSynced: false })
+      this.address = address
+
+      try {
+        this.box = await Box.openBox(address, this.provider)
+        await this._waitForOnSyncDone()
```

Comment:
What would happen if the network was interrupted during this sync I wonder 🤔 

I thought at first that maybe there was an error not being handled somewhere, but... their API doesn't seem to handle this scenario. I guess the only option is to wait, and handle the case where it times out? 🤷‍♂

Reply:
That's a good point. I guess, in the case of a timeout, we should just try again?

Reply:
Sure, that could work. Or just throw an error. Or maybe retry, then throw an error after X failures.

Reply:
- [ ] do something about this

Reply:
After looking over this with @whymarrh , we discovered that the `onSyncDone` function was recently updated to return a promise that correctly handles errors and resolves when the sync is finished. We'll need to update 3Box, but after that we can just await `this.box.onSyncDone()` and it'll "just work". Mostly (the timeout case is still problematic).

The "onSyncDone" from the spaces API is kinda similar - we can await `this.space.syncDone` to achieve the same affect, which is a public but undocumented property. Not ideal, but better than the current solution.

Regarding the timeout, handling that case and throwing an error is simple enough, but we're kinda stuck once the timeout happens. We can't abort the sync, so we can't reliably retry the operation. We'll have to just disable 3Box completely, and not allow the user to enable it (unless they uninstall and reinstall the extension).

If it _does_ return after the timeout, we could allow the user to retry/re-enable it. But if it times out and never returns, we're out of luck.

Reply:
So it looks like the changes to `onSyncDone` to return a promise haven't been released yet. I don't think they need to block the release though, as long as we handle the timeout case. The error will just get caught in handling the timeout. Far from ideal, as we can't recover from it and we force the user to wait, but we can improve things whenever those changes are released.

Reply:
Resolved for now with [47e63ed](https://github.com/MetaMask/metamask-extension/pull/6972/commits/47e63edf9ceedafa7f6f0b2a432d032297aeef7d). Explained here: https://github.com/MetaMask/metamask-extension/pull/6972#issuecomment-522718296